src/VisualStudio/Core/Def/Implementation/Options/LanguageSettingsPersister.cs

```diff
@@ -40,24 +39,37 @@ internal sealed class LanguageSettingsPersister : ForegroundThreadAffinitizedObj
         /// but the ngen image will exist for the basic map between two reference types, since those are reused.</remarks>
         private readonly IBidirectionalMap<string, Tuple<Guid>> _languageMap;
 
+        private IVsTextManager4 _textManager;
+        private ComEventSink _textManagerEvents2Sink;
+
         /// <remarks>
         /// We make sure this code is from the UI by asking for all serializers on the UI thread in <see cref="HACK_AbstractCreateServicesOnUiThread"/>.
         /// </remarks>
         [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
         public LanguageSettingsPersister(
-            [Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider,
-            IGlobalOptionService optionService)
+            IGlobalOptionService optionService, [Import(typeof(SAsyncServiceProvider))] Shell.IAsyncServiceProvider serviceProvider)
             : base(assertIsForeground: true)
         {
-            _textManager = (IVsTextManager4)serviceProvider.GetService(typeof(SVsTextManager));
             _optionService = optionService;
+            _serviceProvider = serviceProvider;
 
             // TODO: make this configurable
             _languageMap = BidirectionalMap<string, Tuple<Guid>>.Empty.Add(LanguageNames.CSharp, Tuple.Create(Guids.CSharpLanguageServiceId))
                                                                .Add(LanguageNames.VisualBasic, Tuple.Create(Guids.VisualBasicLanguageServiceId))
                                                                .Add("TypeScript", Tuple.Create(new Guid("4a0dddb5-7a95-4fbf-97cc-616d07737a77")))
                                                                .Add("F#", Tuple.Create(new Guid("BC6DD5A5-D4D6-4dab-A00D-A51242DBAF1B")))
                                                                .Add("Xaml", Tuple.Create(new Guid("CD53C9A1-6BC2-412B-BE36-CC715ED8DD41")));
+        }
+
+        public async Task InitializeAsync(CancellationToken cancellationToken)
+        {
+            AssertIsForeground();
```

Comment:
This may be what you're alluding to, but just in case: this is a violation of the first threading rule. Async methods should *not* throw if called from an unexpected thread, since they are expected to switch to the thread they require.
[Our VSTHRD109 analyzer](https://github.com/Microsoft/vs-threading/blob/master/doc/analyzers/VSTHRD109.md) helps enforce this, but it does so by recognizing methods that throw. Can you switch this to `ThreadHelper.ThrowIfNotOnUIThread()` and everywhere else in your code so the analyzers can help you?
Otherwise, you can add your own AdditionalFiles to your project that [teach the analyzer about your own helper methods](https://github.com/Microsoft/vs-threading/blob/master/doc/analyzers/configuration.md).